btree/btree.c

```diff
@@ -2928,6 +2928,91 @@ static void btree_callback_credit(struct m0_be_tx_credit *accum)
 {
 	accum->tc_cb_nr += BTREE_CALLBACK_CREDIT;
 }
+
+/**
+ * This function will calculate credits required to allocate node and it will
+ * add those credits to @accum.
+ */
+static void btree_node_alloc_credit(const struct m0_btree  *tree,
+				    struct m0_be_tx_credit *accum)
+{
+	m0_bcount_t             node_size;
+	int                     shift;
+
+	shift     = node_shift(tree->t_desc->t_root);
+	node_size =  1ULL << shift;
+
+	m0_be_allocator_credit(NULL, M0_BAO_ALLOC_ALIGNED,
+			       node_size, shift, accum);
+}
+
+/**
+ * This function will calculate credits required to update @nr nodes and it will
+ * add those credits to @accum.
+ */
+static void btree_node_update_credit(const struct m0_btree  *tree,
+				     struct m0_be_tx_credit *accum,
+				     m0_bcount_t             nr)
+{
+	struct m0_be_tx_credit cred = {};
+ 	m0_bcount_t             node_size;
+	int                     shift;
+
+	shift     = node_shift(tree->t_desc->t_root);
+	node_size =  1ULL << shift;
+
+	cred = M0_BE_TX_CREDIT(1, node_size);
+
+	m0_be_tx_credit_mac(accum, &cred, nr);
+}
+
+/**
+ * This function will calculate credits required to split node and it will add
+ * those credits to @accum.
+ */
+static void btree_node_split_credit(const struct m0_btree  *tree,
+				    struct m0_be_tx_credit *accum)
+{
+	btree_node_alloc_credit(tree, accum);
+	/* credits to update two nodes : existing and newly allocated. */
+	btree_node_update_credit(tree, accum, 2);
+}
+
+/**
+ * This function will calculate credits required for the put KV operation and
+ * add those credits to @accum.
+ */
+static void btree_put_credit(const struct m0_btree  *tree,
+			     struct m0_be_tx_credit *accum)
+{
+	struct m0_be_tx_credit cred = {};
+
+	/* Credits for split operation */
+	btree_node_split_credit(tree, &cred);
+	m0_be_tx_credit_mac(accum, &cred, MAX_TREE_HEIGHT);
+
+	/**
+	 * Additional credits required when extranode is used when height needs
+	 * to be inceased.
+	*/
+	btree_node_alloc_credit(tree, accum);
+	btree_node_update_credit(tree, accum, 1);
```

Comment:
If on line 2992 you have taken care of credits for MAX_TREE_HEIGHT then why are we still worrying about extranode ?
Because during extra node the tree height will be just 2 !!!

Reply:
Yes right. we do not need to allocate credit for extra node, made these changes